ArcGISRuntimeSDKQt_CppSamples/Maps/ReadGeopackage/ReadGeopackage.cpp

```diff
@@ -0,0 +1,125 @@
+  QQuickItem(parent)
+{
+}
+
+void ReadGeopackage::init()
+{
+  // Register the map view for QML
+  qmlRegisterType<MapQuickView>("Esri.Samples", 1, 0, "MapView");
+  qmlRegisterType<ReadGeopackage>("Esri.Samples", 1, 0, "ReadGeopackageSample");
+}
+
+void ReadGeopackage::componentComplete()
+{
+  QQuickItem::componentComplete();
+
+  // find QML MapView component
+  m_mapView = findChild<MapQuickView*>("mapView");
+
+  // Create a map using the topographic basemap
+  m_map = new Map(Basemap::topographic(this), this);
+
+  // Set map to map view
+  m_mapView->setMap(m_map);
+
+  // Create the initial view area
+  Viewpoint initialViewpoint(39.7294, -104.8319, 500000);
+  m_mapView->setViewpoint(initialViewpoint);
+
+  // Initialize the read operation
+  readGeopackage();
+}
+
+// Read the geopackage and create lists for the names and layer data
+void ReadGeopackage::readGeopackage()
+{
+
+  // Load the geopackagea the beginning
+  QString homedir = QDir::homePath();
+  const QString path = homedir + "/ArcGIS/Runtime/Data/gpkg/AuroraCO.gpkg";
+  GeoPackage* auroraGpkg = new GeoPackage(path, this);
+  m_layerList.clear();
+
+  // Make sure there are no errors in loading the geopackage before interacting with it
+  connect(auroraGpkg, &GeoPackage::doneLoading, this, [auroraGpkg, this](Error error)
+  {
+    if (error.isEmpty())
+    {
+
+      // Create two QLists for the different types of layer data within the geopackage (these are combined into m_layerList later on)
+      QList<GeoPackageRaster*> gpkgRasters = auroraGpkg->geoPackageRasters();
+      QList<GeoPackageFeatureTable*> gpkgFeatures = auroraGpkg->geoPackageFeatureTables();
+
+      // Loop through both lists to get the name and layer data. The name data is only used to populate the menu.
+      for (const auto& featureTbl : gpkgFeatures)
+      {
+        QVariantMap layerMap;
+        FeatureLayer* layer = new FeatureLayer(featureTbl, this);
+        layerMap["name"] = layer->name();
+        layerMap["lyr"] = QVariant::fromValue<Layer*>(layer);
+        m_layerList << layerMap;
+      }
+      for (const auto& rasterItm : gpkgRasters)
+      {
+        QVariantMap rasterMap;
+        RasterLayer* rasterLyr = new RasterLayer(rasterItm, this);
+        rasterMap["name"] = rasterLyr->raster()->path().mid(55, 25);
+        rasterMap["lyr"] = QVariant::fromValue<Layer*>(rasterLyr);
+        m_layerList << rasterMap;
+      }
+      emit layerListChanged();
+    }
+  });
+  auroraGpkg->load();
+}
+
+// Getter for the menu names data
+QVariantList ReadGeopackage::layerList() const
+{
+  return m_layerList;
+}
+
+// Called by the QML to toggle visibility at a certain index
+void ReadGeopackage::addOrShowLayer(int index, bool onOff)
+{
+  auto opLayer = m_map->operationalLayers();
+  auto layer = qvariant_cast<Layer*>(m_layerList[index].toMap()["lyr"]);
+  int indexOfLayer = opLayer->indexOf(layer);
+  if(indexOfLayer != -1 && onOff == false)
```

Comment:
`if(` -> `if (`

Reply:
`onOff == false` -> `!visible`

Reply:
What if `onOff` (visible) is `true` here?

Reply:
Sorry I wasn't clear.  There may be some unexpected behavior here in the case of `indexOfLayer != -1` and `onOff == true`.  It would fall into the else and insert another layer which doesn't seem right.

Reply:
@acmorris 
Probably the following would do the trick:
```
if (indexOfLayer != -1)
{
  opLayer->at(indexOfLayer)->setVisible(onOff);
}
else
{
  opLayer->insert(index, layer);
  layer->setVisible(onOff);
}
```